feature/ssl: Added ssl support to apache configuration

cambell-prince opened the pull request:
- Added a new template vhost_ssl.conf.j2 which includes support
  for multiple virtual hosts in the one configuration file.  This allows
  ssl and non ssl to be specified in the one file.
- Copies the ssl key on the remote server.
- Copies the certificate from the local to remote server.
- Ensures the document root exists.
- Changed some variable names to be consistent with ansible convetion.

cambell-prince committed:
feature/ssl: Added ssl support to apache configuration

- Added a new template vhost_ssl.conf.j2 which includes support
  for multiple virtual hosts in the one configuration file.  This allows
  ssl and non ssl to be specified in the one file.
- Copies the ssl key on the remote server.
- Copies the certificate from the local to remote server.
- Ensures the document root exists.
- Changed some variable names to be consistent with ansible convetion.

kosssi commented:
:+1: I test it tomorrow.

Test are failed :( if you can fix it it's the best.

cambell-prince committed:
Fix: Update apt cache to ensure installing latest standard apache

cambell-prince committed:
feature/ssl: Improve tests

- Include 1 test vhost for each template.
- Added key and certificate for testing.

cambell-prince commented:
I've added tests for the ssl feature, and ensured that a vhost is created for each template (there are now two templates).  One thing to note, which isn't great, is that the key copy is from remote to remote (which is good) but there is no idempotent remote copy in stock ansible at this time.  So, I changed the test to reflect that.  If there is an idempotent remote copy implementation that could be used do let me know.

kosssi commented on tasks/vhost.yml:
```diff
@@ -2,9 +2,43 @@
   with_items: apache_vhosts
   when: apache_vhosts != false
+
+- file: path="{{apache_ssl_path}}" state=directory
```
Can you add a name for this task ?

```
- name: Create SSL directory
  file:
    path="{{apache_ssl_path}}"
    state=directory
```

cambell-prince replied:
Done.

kosssi commented on .travis.yml:
```diff
@@ -14,6 +14,6 @@ script:
   - >
     ansible-playbook -i inventory playbook.yml --sudo
-    | grep -q 'changed=0.*unreachable=0.*failed=0'
+    | grep -q 'changed=2.*unreachable=0.*failed=0'
```
It's important for quality of this role which is idempotent.

kosssi commented on tasks/vhost.yml:
```diff
@@ -2,9 +2,43 @@
+  when: apache_vhosts != false
+
+- name: Copy key
+  command: cp "{{apache_ssl_key_store}}/{{item.1.key_file}}" "{{apache_ssl_path}}"
```
why you don't use http://docs.ansible.com/copy_module.html ?

cambell-prince replied:
I don't use the copy module because that doesn't do remote -> remote copies.  It copies local -> remote.  This could be done, but then we have to copy the private key locally, only to copy it back again.  I would rather not do this method.

We could use the synchronize module, however that requires rsync to be installed on the remote, but it is idempotent.  This [Ansible issue](https://github.com/ansible/ansible/issues/9982) suggests using synchronize for remote -> remote copies so I'll switch to that.

kosssi replied:
ah yes it's good idea !!!
Thinks for your contribution.

cambell-prince replied:
I couldn't get the synchronize module to work inside a loop with delegate_to, so I used remote rsync directly and interpret stdout to give idempotence in the copy.

cambell-prince committed:
feature/ssl: Fixed idempotence of ssl key copy

- Added name to ssl folder creation.
- Fixed idempotence of ssl key copy using remote rsync command.

cambell-prince committed:
feature/ssl: Add support for SSLCertificateChainFile

igormukhingmailcom commented:
Cool.
Can you merge it please, @kosssi.

igormukhingmailcom commented:
Thank you :)
